testlapack/dlaqr04.go

```diff
@@ -29,52 +29,73 @@ type dlaqr4Test struct {
 	evWant []complex128 // Optional slice holding known eigenvalues.
 }
 
-func Dlaqr4Test(t *testing.T, impl Dlaqr4er) {
+func Dlaqr04Test(t *testing.T, impl Dlaqr04er) {
 	rnd := rand.New(rand.NewSource(1))
 
-	// Tests that choose the [ilo:ihi+1,ilo:ihi+1] and
-	// [iloz:ihiz+1,ilo:ihi+1] blocks randomly.
+	// Tests for small matrices that choose the ilo,ihi and iloz,ihiz pairs
+	// randomly.
 	for _, wantt := range []bool{true, false} {
 		for _, wantz := range []bool{true, false} {
-			for _, n := range []int{1, 2, 3, 4, 5, 6, 10, 11, 12, 18, 29, 75, 76, 100} {
+			for _, n := range []int{1, 2, 3, 4, 5, 6, 10, 11, 12, 18, 29} {
 				for _, extra := range []int{0, 11} {
-					for cas := 0; cas < min(30, n); cas++ {
-						ilo := rnd.Intn(n)
-						ihi := rnd.Intn(n)
-						if ilo > ihi {
-							ilo, ihi = ihi, ilo
+					for recur := 0; recur <= 2; recur++ {
+						for cas := 0; cas < n; cas++ {
+							ilo := rnd.Intn(n)
+							ihi := rnd.Intn(n)
+							if ilo > ihi {
+								ilo, ihi = ihi, ilo
+							}
+							iloz := rnd.Intn(ilo + 1)
+							ihiz := ihi + rnd.Intn(n-ihi)
+							h := randomHessenberg(n, n+extra, rnd)
+							if ilo-1 >= 0 {
+								h.Data[ilo*h.Stride+ilo-1] = 0
+							}
+							if ihi+1 < n {
+								h.Data[(ihi+1)*h.Stride+ihi] = 0
+							}
+							test := dlaqr04Test{
+								h:     h,
+								ilo:   ilo,
+								ihi:   ihi,
+								iloz:  iloz,
+								ihiz:  ihiz,
+								wantt: wantt,
+								wantz: wantz,
+							}
+							testDlaqr04(t, impl, test, false, recur)
+							testDlaqr04(t, impl, test, true, recur)
 						}
-						iloz := rnd.Intn(ilo + 1)
-						ihiz := ihi + rnd.Intn(n-ihi)
-						h := randomHessenberg(n, n+extra, rnd)
-						if ilo-1 >= 0 {
-							h.Data[ilo*h.Stride+ilo-1] = 0
-						}
-						if ihi+1 < n {
-							h.Data[(ihi+1)*h.Stride+ihi] = 0
-						}
-						test := dlaqr4Test{
-							h:     h,
-							ilo:   ilo,
-							ihi:   ihi,
-							iloz:  iloz,
-							ihiz:  ihiz,
-							wantt: wantt,
-							wantz: wantz,
-						}
-						testDlaqr4(t, impl, test, false)
-						testDlaqr4(t, impl, test, true)
 					}
 				}
 			}
 		}
 	}
+
+	// Tests for matrices large enough to possibly use the recursion (but it
+	// doesn't seem to be the case).
```

Comment:
The recursion takes place only under certain conditions which are not satisfied with the random matrices that are generated. I'll keep this in mind and check the coverage when the whole eigendecomposition solver is finished. Then we'll be able to test it with some standard (hard) test matrices, now we would have to transform them into upper Hessenberg form which would be too troublesome.